packages/js/product-editor/src/components/details-name-block/edit.tsx

```diff
@@ -2,36 +2,175 @@
+	const { editEntityRecord } = useDispatch( 'core' );
+	const productId = useEntityId( 'postType', 'product' );
+	const product: Product = useSelect( ( select ) =>
+		select( 'core' ).getEditedEntityRecord(
+			'postType',
+			'product',
+			productId
+		)
+	);
+
+	const [ name, setName ] = useEntityProp< string >(
+		'postType',
+		'product',
+		'name'
+	);
+
+	const { permalinkPrefix, permalinkSuffix } = useSelect(
+		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+		// @ts-ignore
+		( select: WCDataSelector ) => {
+			const { getPermalinkParts } = select( PRODUCTS_STORE_NAME );
+			if ( productId ) {
+				const parts = getPermalinkParts( productId );
+				return {
+					permalinkPrefix: parts?.prefix,
+					permalinkSuffix: parts?.suffix,
+				};
+			}
+			return {};
+		}
+	);
+
+	const nameIsValid = useValidation(
+		'product/name',
+		() => Boolean( name ) && name !== AUTO_DRAFT_NAME
+	);
 
 	return (
-		<div { ...blockProps }>
-			<TextControl
-				label={ createInterpolateElement(
-					__( 'Name <required />', 'woocommerce' ),
-					{
-						required: (
-							<span className="woocommerce-product-form__optional-input">
-								{ __( '(required)', 'woocommerce' ) }
-							</span>
-						),
-					}
+		<>
+			<div { ...blockProps }>
+				<BaseControl
+					id={ 'product_name' }
+					label={ createInterpolateElement(
+						__( 'Name <required />', 'woocommerce' ),
+						{
+							required: (
+								<span className="woocommerce-product-form__required-input">
+									{ __( '*', 'woocommerce' ) }
+								</span>
+							),
+						}
+					) }
+				>
+					<InputControl
+						name={ 'woocommerce-product-name' }
+						placeholder={ __(
+							'e.g. 12 oz Coffee Mug',
+							'woocommerce'
+						) }
+						onChange={ setName }
+						value={ name || '' }
+					/>
+				</BaseControl>
+				{ productId &&
+					nameIsValid &&
+					product.status === 'publish' &&
+					permalinkPrefix && (
+						<span className="woocommerce-product-form__secondary-text product-details-section__product-link">
+							{ __( 'Product link', 'woocommerce' ) }
+							:&nbsp;
+							<a
+								href={ product.permalink }
+								target="_blank"
+								rel="noreferrer"
+							>
+								{ permalinkPrefix }
+								{ product.slug || cleanForSlug( name ) }
+								{ permalinkSuffix }
+							</a>
+							<Button
+								variant="link"
+								onClick={ () =>
+									setShowProductLinkEditModal( true )
+								}
+							>
+								{ __( 'Edit', 'woocommerce' ) }
+							</Button>
+						</span>
+					) }
+				{ showProductLinkEditModal && (
+					<EditProductLinkModal
+						permalinkPrefix={ permalinkPrefix || '' }
+						permalinkSuffix={ permalinkSuffix || '' }
+						product={ product }
+						onCancel={ () => setShowProductLinkEditModal( false ) }
+						onSaved={ () => setShowProductLinkEditModal( false ) }
+						saveHandler={ async ( updatedSlug ) => {
+							const { slug, permalink } =
+								await updateProductWithStatus(
```

Comment:
Related to the above, do we have to save the product with the new slug right away? or can we just set it as edited state?

Reply:
Likely a design question, but see my [previous comment](https://github.com/woocommerce/woocommerce/pull/37612#discussion_r1163217096).